llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -3246,7 +3246,12 @@ bool SimplifyCFGOpt::SpeculativelyExecuteBB(BranchInst *BI,
 }
 
 /// Return true if we can thread a branch across this block.
-static bool BlockIsSimpleEnoughToThreadThrough(BasicBlock *BB) {
+static bool BlockIsSimpleEnoughToThreadThrough(BasicBlock *BB,
+                                               const TargetTransformInfo &TTI) {
+  // Skip threading if the branch may be divergent.
+  if (TTI.hasBranchDivergence(BB->getParent()))
```

Comment:
Ideally this should be context dependent on uniformity

Reply:
Just another casualty of not being able to update UA! :(

Reply:
yes, if we could query UA for that branch, we could selectively enable this threading if profitable.

Reply:
> Just another casualty of not being able to update UA! :(

Is it possible to update UA on the demand?

Reply:
No it's not. But intuitively it seems that in the case of jump threading, it should be okay to not update UA.

Reply:
The TTI check has nothing to do with the specific block `BB`. A more natural check for this seems to be the outer function `FoldCondBranchOnValueKnownInPredecessor()`, where it can be an early return even before entering the `do-while` loop.